security/authorization/compiler/api/compiler.go

```
@@ -0,0 +1,153 @@
+	}
+	return checked
+}
+func astToCheckedExpr(checked *cel.Ast) *expr.CheckedExpr {
+	checkedExpr, err := cel.AstToCheckedExpr(checked) // v3
+	if err != nil {
+		log.Fatalf("Failed Converting AST to Checked Express %v", err)
+	}
+	return checkedExpr
+}
+
+// CompileYamltoRbac compiles yaml to rbac
+// Must return a pointer to the RBAC proto as to avoid
+// copying the object hence the mutex is not copied
+func CompileYamltoRbac(filename string) *pb.RBAC {
+	yamlFile := ReadYaml(filename)
+	var userPolicy UserPolicy
+	parseYaml(yamlFile, &userPolicy)
+	env := createUserPolicyCelEnv()
+
+	var rbac pb.RBAC
+	rbac.Action = pb.RBAC_Action(pb.RBAC_Action_value[userPolicy.Action])
+	rbac.Policies = make(map[string]*pb.Policy)
+
+	for index := range userPolicy.Rules {
+		rule := userPolicy.Rules[index]
+		name := rule.Name
+		condition := rule.Condition
+		var policy pb.Policy
+		checked := compileCel(env, condition)
+		// checkedExpr := astToCheckedExpr(checked) // v3
+		// policy.CheckedCondition = checkedExpr    // v3
+		checkedExpr := checked.Expr()  // v2
+		policy.Condition = checkedExpr // v2
+		rbac.Policies[name] = &policy
+	}
+	return &rbac
+}
+
+// Converts an expression to a parsed expression, with SourceInfo nil.
+func exprToParsedExpr(condition *expr.Expr) *expr.ParsedExpr {
+	return &expr.ParsedExpr{Expr: condition}
+}
+
+// Converts an expression to a CEL program.
+func exprToProgram(env *cel.Env, condition *expr.Expr) *cel.Program {
+	// ONLY NEEDED FOR V2
+	// v3: can replace line with ast := cel.CheckedExprToAst(checkedExpr)
+	ast := cel.ParsedExprToAst(exprToParsedExpr(condition))
+	program, _ := env.Program(ast)
+	return &program
+}
+
+// Compile takes in input file name and returns serialized output rbac proto
+func Compile(inputFilename string, outputFilename string) {
+	rbac := CompileYamltoRbac(inputFilename)
+	serialized, err := proto.Marshal(rbac)
+	if err != nil {
+		log.Fatalf("Failed to Serialize RBAC Proto %v", err)
```

Let the program panic seems not the right option. You can pass the error out and let the main function decide.

---

So I can pass it to the main but I would still want to fail there right? Since I want to throw the error when the serialization did not work properly